waldo/upload.go

```diff
@@ -80,29 +78,77 @@ func (ua *UploadAction) Perform() error {
 
 	defer ad.Cleanup()
 
-	if err := ua.executeAgent(path, args); err != nil {
+	if err := ua.executeAgent(path, ua.makeAgentArgs()); err != nil {
 		return err
 	}
 
 	if !ciMode && ud != nil && recipe != nil {
-		ua.updateMetadata(ud, recipe, uploadToken)
+		ua.updateMetadata(ud, recipe)
 	}
 
 	return nil
 }
 
 //-----------------------------------------------------------------------------
 
-func (ua *UploadAction) detectAssetVersion() string {
+func (ua *UploadAction) detectAppID(required, allowed bool) (string, error) {
+	appID := ua.options.AppID
+
+	if len(appID) > 0 || required {
+		if !allowed {
+			return "", fmt.Errorf("Option %q not allowed in this context", "--app_id")
+		}
+
+		err := data.ValidateAppID(appID)
+
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return appID, nil
+}
+
+func (ua *UploadAction) detectAppToken(required, allowed bool) (string, error) {
+	appToken := ua.options.UploadToken
+
+	if len(appToken) == 0 {
+		appToken = os.Getenv("WALDO_UPLOAD_TOKEN")
+	}
+
+	if len(appToken) > 0 || required {
+		if !allowed {
+			return "", fmt.Errorf("Option %q not allowed in this context", "--upload_token")
```

Comment:
The value might have come from the environment, not the args

Reply:
Can be a simple change of documentation that the value come from either arg or env with the name, without wiring exactly witch one

Reply:
Fine. I will change the wording of the error.